loader.php

```diff
@@ -15,45 +15,129 @@ function is_development() {
+			}
+		}
+	}
+	$build_assets_path = get_file_path( 'build/asset-manifest.json', $is_theme );
+	if ( file_exists( $build_assets_path ) ) {
+		$build_assets_contents = file_get_contents( $build_assets_path );
+		if ( ! empty( $build_assets_contents ) ) {
+			// Prepend "build/" to all build-directory array paths.
+			return array_map(
+				function( $asset_path ) { return 'build/' . $asset_path; },
+				array_values( json_decode( $build_assets_contents, true ) )
+			);
+		}
 	}
+	return null;
+}
+
 
-	$path = get_theme_file_path( 'react-port' );
-	if ( ! file_exists( $path ) ) {
-		return null;
+/**
+ * Return web URIs or convert relative filesystem paths to absolute paths.
+ *
+ * @param string $asset_path  A relative filesystem path or resource URI.
+ * @param bool   $is_theme    Whether to treat the $path as a theme file.
+ * @return string
+ */
+function get_asset_uri( string $asset_path, bool $is_theme ) {
+	if ( strpos( $asset_path, '://' ) !== false ) {
+		return $asset_path;
+	}
+	if ( ! is_development() ) {
+		$asset_path = 'build/' . $asset_path;
 	}
+	return $is_theme
+		? get_theme_file_uri( $asset_path )
+		: plugin_dir_url( __file__ ) . $asset_path;
+}
 
-	$port = file_get_contents( $path );
-	if ( empty( $port ) || ! is_numeric( $port ) ) {
-		return null;
+/**
+ * Search for webpack-manifest-plugin output, attempt to load it, and enqueue
+ * any scripts or styles contained within.
+ *
+ * @param bool $is_theme Whether to look for files in a theme or plugin context.
+ */
+function autoenqueue_assets( bool $is_theme ) {
+	// Attempt to load
+	$assets = get_assets_list( $is_theme );
+	if ( empty( $assets ) ) {
+		// TODO: This should be an error condition.
+		return;
 	}
 
-	return (int) trim( $port );
-}
-
-function enqueue_assets( $id, $deps = [] ) {
-	$port = get_react_port();
-	if ( $port ) {
-		wp_enqueue_script(
-			$id,
-			sprintf( 'http://localhost:%d/static/js/bundle.js', $port ),
-			$deps,
-			null,
-			true
-		);
-	} else {
-		wp_enqueue_script(
-			$id,
-			get_theme_file_uri( 'build/js/main.js' ),
-			$deps,
-			null,
-			true
-		);
+	foreach ( $assets as $asset_path ) {
+		$is_js = preg_match( '/\.js$/', $asset_path );
+		$is_css = preg_match( '/\.css$/', $asset_path );
+		if ( ! $is_js && ! $is_css ) {
+			continue;
+		}
+		// Asset file contains URIs, enqueue them directly.
+		if ( $is_js ) {
+			wp_enqueue_script(
+				get_asset_key( $asset_path ),
+				get_asset_uri( $asset_path, $is_theme ),
+				[],
+				null,
+				true
+			);
+		} else if ( $is_css ) {
+			wp_enqueue_style(
+				get_asset_key( $asset_path ),
+				get_asset_uri( $asset_path, $is_theme )
+			);
+		}
 	}
 }
+
+/**
+ * Wrapper function to auto-enqueue all built assets for a theme.
+ */
+function autoenqueue_theme_assets() {
+	autoenqueue_assets( true );
+}
+
+/**
+ * Wrapper function to auto-enqueue all built assets for a plugin.
+ */
+function autoenqueue_plugin_assets() {
+	autoenqueue_assets( false );
+}
```

Comment:
It would be great to have a universal way of getting the path or URI for a file regardless of whether it's in a theme or a plugin, but I am not aware of one. If there's a good option out there (or a better solution with what we have, if not) I'd welcome the input, I don't much like passing boolean flags all over the place :)

Reply:
Had not seen #1, I can adapt that interface